### PR TITLE
Replace pkg-dir with custom lookup

### PR DESCRIPTION
**Pull request by TrySound**
I want my node_modules to be small.

**What kind of change does this PR introduce?**

Replaced pkg-dir with custom lookup and got rid from a few transitive dependencies.

**Did you add tests for your changes?**

Already covered by tests

**Does this PR introduce a breaking change?**

No

**What needs to be documented once your changes are merged?**

Nothing

### CODE REVIEW

**Review by alexander-akait — APPROVED** [review]
Great, thanks

**Comment by TrySound** [comment]
Not sure what to do with this error. Any idea?

```
lib/config/defaults.js(267,21): error TS2349: This expression is not callable.
  Type 'typeof import("/home/vsts/work/1/s/node_modules/escalade/sync/index")' has no call signatures.
```

**Comment by alexander-akait** [comment]
Maybe we need update `ts`, because we use `beta` (https://github.com/webpack/webpack/blob/master/package.json#L98) due some problems with types for `4.1`

**Comment by TrySound** [comment]
I guess the problem is that type annotation uses `export default` which ts consider as `exports.default = `. Is it ok to suppress this error?

**Comment by alexander-akait** [comment]
Ideally it should not have the problem :smile: Need look why it happens, maybe we should open an issue in `escalade` repo if types wrong

**Comment by TrySound** [comment]
I guess using ts with commonjs is very rare use case as nobody complained before.

**Comment by TrySound** [comment]
Won't fix. TS does not let to solve this without breaking change. Too much fragmentation in configuration.

**Comment by alexander-akait** [comment]
Let's use `@ts-ignore` with comments why we can't solve it

**Comment by TrySound** [comment]
@alexander-akait Does it look fine now?

**Comment by alexander-akait** [comment]
I think yes

File: lib/config/defaults.js [hunk 1]
```diff
@@ -262,9 +262,17 @@ const applyCacheDefaults = (cache, { name, mode }) => {
...
-				const dir = pkgDir.sync(cwd);
+				// eslint-disable-next-line no-warning-comments
+				// @ts-ignore: escalade typings use es module "export default" which is treated as "{ default: escalade }" by typescript; can be fixed with migration to es modules
+				const pkgPath = escalade(cwd, (dir, names) => {
+					if (names.includes("package.json")) {
+						return "package.json";
+					}
+					return false;
+				});
+				const dir = path.dirname(pkgPath);
```
Comment (sokra) [inline lib/config/defaults.js:275]: If you want it to be even smaller, you could use this piece of code instead:

``` js
let dir = process.cwd();
for (;;) {
  try {
    if (fs.statSync(path.join(dir, "package.json").isFile()) break;
  } catch (e) {}
  const parent = path.dirname(dir);
  if (dir !== parent) { dir = undefined; break; }
}
```

**Review by sokra — CHANGES_REQUESTED** [review]

File: lib/config/defaults.js (hunk 1 above)
Comment (TrySound) [reply lib/config/defaults.js]: Done

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**